### PR TITLE
Return rendered value and renderer for syntax highlighting in UI.

### LINKED ISSUE #39527: Add `json` and `sql` template rendering to Grid View
### Apache Airflow version

2.9.1

### If "Other Airflow 2 version" selected, which one?

_No response_

### What happened?

In Airflow v2.9.1, the `Rendered Template` link is now missing from the `Details` tab when in `Grid` view.

### 2.8.4
Screenshot 2024-05-09 at 16 18 01

### 2.9.1
Screenshot 2024-05-09 at 16 18 53

### What you think should happen instead?

It is still possible to get to the rendered template via `More Details` but having this link here is much more convenient. 

### How to reproduce

Simply start an instance of Airflow 2.9.1.

### Operating System

n/a

### Versions of Apache Airflow Providers

n/a

### Deployment

Astronomer

### Deployment details

_No response_

### Anything else?

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by tirkarthi**
closes: #39527
related: #39527

This PR is like POC of the approach. The main point is to return the renderer information for the template field to UI and let react syntax highlighter module use the info to apply appropriate language syntax highlighting. There are four changes

1. To get the renderers the task needs to be retrieved from dagbag which in turn needs the existing session attached to be passed to get_dag or else the new session created will commit and cause the one attached to task_instance during serialization to become stale.
2. Earlier the rendered_fields will return only the value and now it will return {"value": value, "renderer": renderer} which though defined as a dictionary without structure might break scripts using it. 
3. Since configuration.query.query which is also a rendered field previously it was not returned in API but visible in the legacy page. Such nested paths are also now returned in the API.
4. Rendered template fields should be rendered as it is and not converted to camelcase which is also fixed here by excluding nested keys when rendered_fields key is present in response.

Thoughts welcome on the change and possible approaches.

Before PR : 

```
  "rendered_fields": {
    "configuration": {
        "query": {
          "query": "INSERT INTO `mylong_table_name`\n            (id,\n             name,\n             age)\nSELECT id,\n       name,\n       age\nFROM   users;"
        }
    },
    "project_id": 1
  },
```

With PR : 

```
  "rendered_fields": {
    "configuration": {
      "renderer": "json",
      "value": {
        "query": {
          "query": "INSERT INTO `mylong_table_name`\n            (id,\n             name,\n             age)\nSELECT id,\n       name,\n       age\nFROM   users;"
        }
      }
    },
    "configuration.query.query": {
      "renderer": "sql",
      "value": "INSERT INTO `mylong_table_name`\n            (id,\n             name,\n             age)\nSELECT id,\n       name,\n       age\nFROM   users;"
    },
    "project_id": {
      "renderer": null,
      "value": 1
    }
  },
```
Screenshot : 

image

### CODE REVIEW

**Comment by bbovenzi** [comment]
Nice! In the example screenshots, why would we want to show the `configuration` and then `configuration.query.query`? It seems like the second one is sufficient in this case.

**Comment by tirkarthi** [comment]
yes, the case is taken from the linked issue but there could be other cases where `configuration` could have some keys that are not templated but are useful along with the overall structure being useful to debug in case of complex json. The change also means changing the API. Right now API only returns `configuration` value which is part of `template_fields` and doesn't return `configuration.query.query` which is a nested value part of `template_fields_renderers` mapping. The legacy view takes template_fields and also renders the fields from `template_fields_renderers` as separate fields.

Since the PR has test case failure I was thinking to keep `rendered_fields` as it is with nested values enriched and have a new field for the renderer mapping. Something like below structure so that it doesn't break current API and also cases where the json might have field `renderer` which will get overridden.

Sample response :

```json
  "rendered_fields": {
    "configuration": {
        "query": {
          "query": "INSERT INTO `mylong_table_name`\n            (id,\n             name,\n             age)\nSELECT id,\n       name,\n       age\nFROM   users;"
        }
    },
    "configuration.query.query": "INSERT INTO `mylong_table_name`\n            (id,\n             name,\n             age)\nSELECT id,\n       name,\n       age\nFROM   users;",
    "project_id": 1
  },
  "rendered_fields_mapping": {
    "configuration": "json",
    "configuration.query.query": "sql"
  },
```

Sample operator

```python
class CustomOperator(BaseOperator):

    template_fields: list[str] = (
        "configuration",
        "project_id",
    )
    template_ext: list[str] = (
        ".json",
        ".sql",
    )
    template_fields_renderers = {
        "configuration": "json",
        "configuration.query.query": "sql",
    }

    def __init__(self, configuration, project_id, **kwargs) -> None:
        super().__init__(**kwargs)
        self.configuration = configuration
        self.project_id = project_id

    def execute(self, context):
        print(self.configuration)
        return self.configuration

```

**Comment by nathadfield** [comment]
> Nice! In the example screenshots, why would we want to show the `configuration` and then `configuration.query.query`? It seems like the second one is sufficient in this case.

The `configuration` object might contain all sorts of relevant templated information that relates to a BigQuery job besides just the SQL. https://cloud.google.com/bigquery/docs/reference/rest/v2/Job#JobConfigurationQuery

**Comment by tirkarthi** [comment]
Since the change in response will break API compatibility I have kept the original response for `rendered_fields` and added a new field `template_fields_renderers` which just returns the same attribute set on the task as a dictionary of mapping to highlighter. Then in the frontend for each key in the `rendered_fields` the corresponding highlighter will be looked up in `template_fields_renderers` to highlight the value. 

```json
  "rendered_fields": {
    "configuration": {
      "query": {
        "query": "INSERT INTO `mylong_table_name`\n            (id,\n             name,\n             age)\nSELECT id,\n       name,\n       age\nFROM   users;"
      }
    },
    "configuration.query.query": "INSERT INTO `mylong_table_name`\n            (id,\n             name,\n             age)\nSELECT id,\n       name,\n       age\nFROM   users;",
    "project_id": 1
  },

  "template_fields_renderers": {
    "configuration": "json",
    "configuration.query.query": "sql"
  },
```

**Comment by tirkarthi** [comment]
This still needs test case for new attribute `template_fields_renderers` and the nested fields being returned in the response along with frontend code cleanup but will be happy to know thoughts on the approach.

Thanks.

File: airflow/www/static/js/dag/details/taskInstance/Details.tsx [hunk 1]
```diff
@@ -294,22 +309,51 @@ const Details = ({ gridInstance, taskInstance, group }: Props) => {
...
                   }
+
                   return (
                     <Tr key={key}>
                       <Td>{key}</Td>
                       <Td>
-                        <Code fontSize="md">{field as string}</Code>
+                        <Flex alignItems="right">
+                          <SyntaxHighlighter
+                            fontSize="md"
```
Comment (bbovenzi) [inline airflow/www/static/js/dag/details/taskInstance/Details.tsx]: Let's fix this style with something like

```
const theme = useTheme();
```

and

```
                            customStyle={{
                              fontSize: theme.fontSizes.md,
                              font: theme.fonts.mono,
                            }}
```

**Comment by bbovenzi** [comment]
Let's rebase but otherwise this looks like an improvement to me and we can continue to iterate on it.

**Comment by tirkarthi** [comment]
Thanks @bbovenzi , I have rebased and fixed conflicts with changes made to use RenderedJsonField in another PR. One issue I am facing is as per below screenshot I have project_id with value of 1. According to `JSON.parse` in Firefox it's valid. The code used to check for valid JSON in RenderedJsonField.tsx also returns true but the src value is not accepted by ReactJson which is strange.


image

**Comment by tirkarthi** [comment]
I also wish sometime in Airflow 3 or later all the metadata related to a task like `template_fields_renderers` are available without parsing the DAG in API server which would have been helpful in this case.

**Comment by bbovenzi** [comment]
Oh maybe we need to add a check to RenderedJson to just render simple variables (ex: number, string, boolean) as-is vs as json.

Also looks like `tests/api_connexion/schemas/test_task_instance_schema.py::TestTaskInstanceSchema::test_task_instance_schema` with or without sla tests are failing.

**Comment by cah-jonathan-cachat01** [comment]
When will we get our Rendered Template page back?

**Comment by nathadfield** [comment]
@tirkarthi I am keeping this PR alive.  What's the status with this?

**Comment by tirkarthi** [comment]
> @tirkarthi I am keeping this PR alive. What's the status with this?

Sorry, with Airflow 3 and legacy UI deprecated I am afraid I will not be able to get back to this. I will be happy to do this change in the new UI once the design and API changes for this is in place.

**Comment by potiuk** [comment]
> > @tirkarthi I am keeping this PR alive. What's the status with this?
> 
> Sorry, with Airflow 3 and legacy UI deprecated I am afraid I will not be able to get back to this. I will be happy to do this change in the new UI once the design and API changes for this is in place.

+10 on that

**Pull request closed without merging**